### DISCUSSION TITLE
Finding and Deleting an edge

### DISCUSSION THREAD

**Discussion by benanavd** [ArcadeData/arcadedb#258] [Q&A] [answered]
My question concerns deleting an edge. Suppose I have the RID of the head vertex and the RID of the tail vertex and the name of the EdgeType.  What is the best way to find and then delete the edge in Java code and what is the complexity of doing so? Does it have to iterator over all the edges connected to the head vertex to find it?

**Comment by lvca** [comment, marked as answer]
@benanavd I'd say that is the least efficient way to find an edge in a graph database :-) In that case, you're right, you should start from the head or tail and look for the edge. The best would be having its RID so the complexity would be `O(log1)`. In your case would be `O(log1)` to look up for the head or tail vertex + browse the edges. Then, the edges in the list are not sorted, so it depends on how many edges you have.

A solution would be to create an index on the edge on `@out` and `@in`, so you can use the index to look up the edge quickly.

Example:

```
CREATE INDEX Watched_out_in ON Watched (`@out`, `@in`) UNIQUE
```

You can omit UNIQUE if you can have multiple edges between the same vertices. Creating a unique index on edge out+in is useful to guarantee you have only one edge, so it works as a constraint. The only cons of having an index on edge's out+in is the insertion will be slower, because an index must be updated every time.

**Reply by benanavd** [reply]
For more clarity, can you provide the java code?  So from your documentation  to make elon a friend of steve:

    try( Database db = databaseFactory.open(); ){
      db.transaction( () -> {
      MutableVertex elon = db.newVertex("User", "name", "Elon", "lastName", "Musk");
      MutableVertex steve = db.newVertex("User", "name", "Steve", "lastName", "Jobs");
     elon.newEdge("IsFriendOf", steve, true, "since", 2010);
     });
    }

Now somewhere later in code, elon is no longer a friend of steve (and I don't have the edge RID). How would I do that in the code. Also what is the code to create  the appropriate index?  Since friends is not single valued, the index cannot be unique.

Thanks!

**Reply by lvca** [reply]
You can execute this first:

```java
db.command( "sql", "CREATE INDEX IsFriendOf_out_in ON IsFriendOf (`@out`, `@in`) NOTUNIQUE");
```

And then you can delete the edge with something like this (passing elonRID and steveRID):

```java
db.command( "sql", "DELETE FROM ( SELECT FROM IsFriendOf WHERE `@out` = ? AND `@in` = ? )", elonRID, steveRID );
```

**Reply by benanavd** [reply]
OK. Thank you. I was hoping to be able to do something like (assuming I have references to elon and steve):
    `MutableVertex elon, steve;
     elon.deleteEdge("IsFriendOf", steve);`

**Reply by lvca** [reply]
You could have multiple edges between elon and steve, so an API like that could be misleading.

**Reply by benanavd** [reply]
I was thinking of IsFriendOf as having no properties. I believe that is what you call a light edge? In any case, I get your point. It's probably better to do things with queries as you suggest because that way presumably that way you can optimized how the query solution is determined.

From the documentation, I don't see anything about creating indexes as you suggest. Maybe I missed it somewhere?  Seems like it might be important to be able to create such indexes in certain cases?

**Reply by lvca** [reply]
Correct, that could be a light edge. You can enable a special optimization for managing light edges, but I won't suggest it unless you need extreme tuning after having tried everything else.

In the docs, that index is in an example to create a constraint: https://docs.arcadedb.com/#SQL-Create-Edge
